addons/survey/models/survey_question.py

```diff
@@ -152,173 +152,112 @@ def _onchange_is_page(self):
-            errors.update({answer_tag: self.constr_error_msg})
-        # Checks if user input is a number
-        if answer:
-            try:
-                floatanswer = float(answer)
-            except ValueError:
-                errors.update({answer_tag: _('This is not a number')})
-        # Answer validation (if properly defined)
-        if answer and self.validation_required:
+    def _validate_numerical_box(self, answer):
+        try:
+            floatanswer = float(answer)
+        except ValueError:
+            return {self.id: _('This is not a number')}
+
+        if self.validation_required:
             # Answer is not in the right range
             with tools.ignore(Exception):
-                floatanswer = float(answer)  # check that it is a float has been done hereunder
                 if not (self.validation_min_float_value <= floatanswer <= self.validation_max_float_value):
-                    errors.update({answer_tag: self.validation_error_msg})
-        return errors
+                    return {self.id: self.validation_error_msg}
+        return {}
 
-    def date_validation(self, date_type, post, answer_tag, min_value, max_value):
-        self.ensure_one()
-        errors = {}
-        if date_type not in ('date', 'datetime'):
-            raise ValueError("Unexpected date type value")
-        answer = post[answer_tag].strip()
-        # Empty answer to mandatory question
-        if self.constr_mandatory and not answer:
-            errors.update({answer_tag: self.constr_error_msg})
+    def _validate_date(self, answer):
         # Checks if user input is a date
-        if answer:
-            try:
-                if date_type == 'datetime':
-                    dateanswer = fields.Datetime.from_string(answer)
-                else:
-                    dateanswer = fields.Date.from_string(answer)
-            except ValueError:
-                errors.update({answer_tag: _('This is not a date')})
-                return errors
+        try:
+            if self.question_type == 'datetime':
+                dateanswer = fields.Datetime.from_string(answer)
+            else:
+                dateanswer = fields.Date.from_string(answer)
+        except ValueError:
+            return {self.id: _('This is not a date')}
         # Answer validation (if properly defined)
-        if answer and self.validation_required:
+        if self.validation_required:
             # Answer is not in the right range
             try:
-                if date_type == 'datetime':
+                if self.question_type == 'datetime':
                     date_from_string = fields.Datetime.from_string
                 else:
                     date_from_string = fields.Date.from_string
                 dateanswer = date_from_string(answer)
-                min_date = date_from_string(min_value)
-                max_date = date_from_string(max_value)
+                min_date = date_from_string(self.validation_min_date)
```

Comment:
Why do we need to use date_from_string on actual fields.Date values?

Also, don't you need to use different fields based on the type of the question:

```
validation_min_date = fields.Date('Minimum Date')
validation_max_date = fields.Date('Maximum Date')
validation_min_datetime = fields.Datetime('Minimum Datetime')
validation_max_datetime = fields.Datetime('Maximum Datetime')
```